source/distributed/tria.cc

```diff
@@ -3436,10 +3597,30 @@ namespace parallel
 
       // partition the new mesh between all processors
       if (!(settings & no_automatic_repartitioning))
-        dealii::internal::p4est::functions<dim>::
-        partition (parallel_forest,
-                   /* prepare coarsening */ 1,
-                   /* weight_callback */ NULL);
+        {
+          // copy the cell weights into the order in which p4est will
+          // encounter them, then attach (temporarily) a pointer to
+          // this list through p4est's user_pointer object
+          Assert (parallel_forest->user_pointer == this,
+                  ExcInternalError());
+          PartitionWeights<dim,spacedim> partition_weights (*this,
```

Comment:
Yes, the code you looked at was intended to just ensure that we get the same results if all cells have weight one, compared to the case where we provide no weight at all. This verified, I broke the code out as above.